tests/unit/frameworks/dynamic_sparse/conv_coactivations_test.py

```diff
@@ -0,0 +1,176 @@
+    i2 -= p2
+
+    i1 += n * s1
+    i2 += m * s2
+
+    indxs = []
+    for c_in in range(in_channels):
+        for n_k1 in range(k1):
+            for m_k2 in range(k2):
+                filter_indx = (c_in, n_k1, m_k2)
+                input_indx = (c_in, i1 + n_k1, i2 + m_k2)
+                indxs.append((input_indx, filter_indx))
+
+    return indxs
+
+
+def calc_coactivations(
+    shape,
+    padding,
+    stride,
+    input_tensor,
+    output_tensor,
+    mean_activations,
+    alpha
+):
+
+    c_out = shape[0]
+    c_in = shape[1]
+    kernel_size = (shape[2], shape[3])
+
+    b_out = output_tensor.shape[0]
+    n_out = output_tensor.shape[2]
+    m_out = output_tensor.shape[3]
+
+    h = torch.zeros(shape)
+    for b_ in range(b_out):
+        for c_out_ in range(c_out):
+            for n_out_ in range(n_out):
+                for m_out_ in range(m_out):
+                    unit_1 = output_tensor[b_, c_out_, n_out_, m_out_]
+                    indxs = get_indeces_of_input_and_filter(
+                        n_out_, m_out_, c_in, kernel_size, padding, stride)
+
+                    for input_indx, filter_indx in indxs:
+                        c_in_, n_in_, m_in_ = input_indx
+                        c_fl_, n_fl_, m_fl_ = filter_indx
+                        unit_2 = input_tensor[b_, c_in_, n_in_, m_in_]
+
+                        if coactivation(unit_2,
+                                        unit_1, alpha, mean_activations):
+                            h[c_out_, c_fl_, n_fl_, m_fl_] += 1
+    return h
+
+
+# ------------------
+# Tests
+# ------------------
+
+
+class CoactivationsTest(unittest.TestCase):
+    def setUp(self):
+        pass
+
+    def test_via_brute_force_comparision(self):
+
+        batch_size = 2
+        in_channels = 4
+        out_channels = 4
+        kernel_size = (2, 2)
+        stride = (1, 1)
+        padding = 0
+
+        conv = DSConv2d(
+            in_channels=in_channels,
+            out_channels=out_channels,
+            kernel_size=kernel_size,
+            stride=stride,
+            padding=padding,
+            dilation=1,
+            groups=1,
+            sparsity=0.98,
+            hebbian_prune_frac=0.99,
+            prune_dims=[],
+            magnitude_prune_frac=0.00,
+            coactivation_test="variance",
+            update_nsteps=1,
+        )
+        conv.init_coactivation_tracking()
+
+        input_tensor = torch.randn(batch_size, in_channels, *kernel_size)
+        output_tensor = conv(input_tensor)
+        mean_activations = (input_tensor.mean(), output_tensor.mean())
+
+        kernel_size = conv.kernel_size
+        stride = conv.stride
+        padding = conv.padding
+        alpha = conv.get_activity_threshold(input_tensor, output_tensor)
+
+        coacts = calc_coactivations(
+            conv.weight.shape, padding, stride,
+            input_tensor, output_tensor, mean_activations, alpha)
+        assert conv.coactivations.allclose(coacts, atol=0, rtol=0)
```

Comment:
In unit tests, should use self.assertTrue or other unit test comparisons.  See https://docs.python.org/3/library/unittest.html#assert-methods

Reply:
Fixed